File: backend/app/schemas/place.py
import uuid
from datetime import datetime

from pydantic import BaseModel

from app.models.place import Category, Source


class PlaceSearchResult(BaseModel):
    external_id: str
    source: Source
    name: str
    lat: float
    lng: float
    category: Category
    address: str
    phone: str | None = None
    website: str | None = None
    rating: float | None = None


class AddPlaceRequest(BaseModel):
    source: Source
    external_id: str


class PlaceResponse(BaseModel):
    id: uuid.UUID
    name: str
    lat: float
    lng: float
    category: Category
    address: str
    phone: str | None
    website: str | None
    source: Source
    external_id: str
    created_at: datetime

    model_config = {"from_attributes": True}


class UserPlaceResponse(BaseModel):
    id: uuid.UUID
    place: PlaceResponse
    rating: float | None
    notes: str | None
    visited: bool
    want_to_visit: bool
    created_at: datetime

    model_config = {"from_attributes": True}


class RatePlaceRequest(BaseModel):
    rating: float
    notes: str | None = None
    visited: bool = True

File: ios/NightOut/Core/Models/Place.swift
import Foundation

enum PlaceCategory: String, Codable {
    case bar, restaurant, cafe
}

enum PlaceSource: String, Codable {
    case google, foursquare, osm
}

struct Place: Codable, Identifiable {
    let id: UUID
    let name: String
    let lat: Double
    let lng: Double
    let category: PlaceCategory
    let address: String
    let phone: String?
    let website: String?
    let source: PlaceSource
    let externalId: String
    let createdAt: Date

    enum CodingKeys: String, CodingKey {
        case id, name, lat, lng, category, address, phone, website, source
        case externalId = "external_id"
        case createdAt = "created_at"
    }
}

struct UserPlace: Codable, Identifiable {
    let id: UUID
    let place: Place
    var rating: Double?
    var notes: String?
    var visited: Bool
    var wantToVisit: Bool
    let createdAt: Date

    enum CodingKeys: String, CodingKey {
        case id, place, rating, notes, visited
        case wantToVisit = "want_to_visit"
        case createdAt = "created_at"
    }
}

struct PlaceSearchResult: Codable, Identifiable {
    var id: String { "\(source.rawValue):\(externalId)" }
    let externalId: String
    let source: PlaceSource
    let name: String
    let lat: Double
    let lng: Double
    let category: PlaceCategory
    let address: String
    let phone: String?
    let website: String?
    let rating: Double?

    enum CodingKeys: String, CodingKey {
        case source, name, lat, lng, category, address, phone, website, rating
        case externalId = "external_id"
    }
}
